[mono] Implement AdvSimd

imhameed opened the pull request:
This change adds AdvSimd and AdvSimd.Arm64 support to LLVM-enabled Mono.

Most aarch64 LLVM intrinsic functions are overloaded and have names determined
by an invariant base string prepended to a string representation of one or two
type parameters. Intrinsic functions used by an LLVM module must have a
declaration somewhere in memory when JITting or somewhere in the output bitcode
file when AOTing. Currently Mono maintains a hash table that maps internal
intrinsic IDs to LLVM intrinsic declarations. These IDs have been extended: a
simplified type representation is added to the key's upper bits. This
representation is not especially compact, and currently uses 9 bits to label 18
states, but it's easy to look at in a debugger. (A simple base-18 encoding
could encode three parameters in 13 bits.)

These overload-tagged IDs can be passed to
`OP_XOP_OVR{_,_SCALAR,_BYSCALAR}X_{X,X_X,X_X_X}`. The return type of the
intrinsic that generates these mini ops is used to derive the overload tag to
find the corresponding LLVM intrinsic function declaration.

`MonoLLVMModule::intrins_by_id` is removed, because LLVM intrinsic lookup keys
are no longer small contiguous integers. It only seemed to serve as a lookup
table for data already contained in a hash table.

The corresponding instructions for some of these .NET-level intrinsics take
immediate parameters. For some of these instructions, the LLVM IR code that
selects these immediate-argument instructions can emit a fallback for
non-constant parameters, either by using an equivalent instruction with a
register operand or by using a longer and less-efficient instruction sequence.
For the rest, a branching code sequence is emitted. Helper functions
(`immediate_unroll_begin` etc.) are added to make this a little less
repetitious.

Some operations take an immediate operand denoting a lane to select in a vector
before proceeding with another generic vector or scalar operation. These are
decomposed into a sequence of `OP_ARM64_SELECT_SCALAR` followed by the
non-lane-specific operation. LLVM can still optimize this to the lane-selecting
instruction when possible, and can generate fallback code for non-immediate
lane selection.

The tables describing the intrinsics supported by the runtime are extended to
support intrinsics with different target instructions for signed, unsigned and
floating point parameters. Whenever possible, .NET-level intrinsics that
correspond to a single LLVM intrinsic function are stored as a single entry in
these tables. Unfortunately many intrinsics need to be translated into a
sequence of LLVM IR operations; for these, new mini IR opcodes are added to
select the LLVM IR builder code that should run.

imhameed committed:
Checkpoint

(Insert meaningful description here)

imhameed committed:
Add some shifts

imhameed committed:
Implement rounding

imhameed committed:
Implement ReverseElement{Bits,8,16,32}

imhameed committed:
Add reciprocal fp/u32 operations

imhameed committed:
Add some bitwise operations

imhameed committed:
Add negation

imhameed committed:
Add every AdvSimd symbol name

imhameed committed:
Minor cleanup

imhameed committed:
Checkpoint

Remove `MonoLLVMModule::intrins_by_id`, which doesn't do anything other
than serve as a lookup table for data contained in `intrins_id_to_intrins`

Don't emit table-driven intrinsics when the corresponding intrinsic
group isn't fully supported.

imhameed committed:
Implement some multiplication ops

imhameed committed:
Fix amd64

imhameed committed:
More multiplication

imhameed committed:
Implement min/max

imhameed committed:
Add load ops

imhameed committed:
Add clz/cls

imhameed committed:
Implement insert/insertscalar

imhameed committed:
Add fused ops

imhameed committed:
Add extract

imhameed committed:
Add duplicate

imhameed committed:
Add divide

imhameed committed:
Add f32->{u32,i32}

imhameed committed:
Add fp64->{u64,i64}

imhameed committed:
arm64 fp conversions

imhameed committed:
Add comparisons

imhameed committed:
Implement ceiling

imhameed committed:
Revert some stray unrelated changes

imhameed committed:
Implement more bitwise operations

imhameed committed:
More arithmetic

imhameed committed:
Implement Abs.*, Absolute.*

imhameed committed:
Implement AbsoluteCompare.*

imhameed committed:
Misc fixes to shifts, math

imhameed committed:
Implement missing stuff from AdvSimd.Arm64

imhameed committed:
Implement the AdvSimd.Arm64 parts of fma

imhameed committed:
More AdvSimd.Arm64

imhameed committed:
Implement ExtractVector64/ExtractVector128

imhameed committed:
Unroll instructions with immediate constants and no fallback support via LLVM

imhameed commented:
There are some compilation failures in the test cross-compilation step on CI. They are related, and I'm debugging them right now.

imhameed committed:
... Fix MinPairwiseScalar/MinAcross/MaxPairwiseScalar/MaxAcross

imhameed committed:
Fix brain-os for ld1/st1

imhameed committed:
???

vargaz commented on src/mono/mono/mini/mini-llvm-cpp.h:
```diff
@@ -16,22 +16,14 @@
 #include "llvm-c/ExecutionEngine.h"
 
+#include "mini-llvm.h"
+
```
We didn't include mono headers into this file because it includes mono's config.h etc. which used to confuse the build of mini-llvm-cpp.cpp.

imhameed replied:
Ah. I might just add another file that only contains the enum definition for IntrinsicId then. Or maybe there's an existing header that'd be a more appropriate place to put this (I haven't looked very hard).

vargaz replied:
That would be the best approach imho, we already have a bunch of headers with enums like patch-info.h, wrapper-types.h etc.

imhameed commented on src/mono/mono/mini/mini.h:
```diff
@@ -2877,24 +2877,6 @@ enum {
-	SIMD_OP_ARM64_DABSOLUTE_COMPARE_LESS_THAN,
-	SIMD_OP_ARM64_FABSOLUTE_COMPARE_LESS_THAN_OR_EQUAL,
-	SIMD_OP_ARM64_DABSOLUTE_COMPARE_LESS_THAN_OR_EQUAL,
 	SIMD_OP_SSE_CVTSS2SI,
```
Yeah.

imhameed commented:
Test cross-compilation succeeded, but there are test failures on the arm64 LLVM Helix lanes that are related. I am debugging these now.

imhameed committed:
Fix floating point Add/Subtract

imhameed committed:
Fix CompareEqualScalar/CompareGreaterThanOrEqualScalar/CompareGreaterThanScalar

imhameed committed:
Fix ShiftRightArithmeticRounded, ShiftRightArithmeticRoundedScalar, ShiftRightArithmeticRoundedAddScalar

imhameed committed:
Fix MultiplyDoublingWideningLowerAndSubtractSaturate, FusedMultiplySubtractBySelectedScalar, FusedMultiplySubtractByScalar

imhameed committed:
Fix ShiftLogicalSaturateScalar, ShiftArithmeticRoundedSaturateScalar, ShiftArithmeticSaturateScalar, ShiftLeftLogicalSaturate and ShiftLeftLogicalSaturateScalar

Fix ShiftLeftLogicalSaturate and ShiftLeftLogicalSaturateScalar:
decompose it into a promotion of the second argument into a vector
followed by an overloaded invocation of @llvm.aarch64.neon.uqshl or
@llvm.aarch64.neon.sqshl

imhameed committed:
Fix ShiftLeftLogicalSaturateUnsignedScalar, ShiftLogicalRoundedSaturateScalar

ShiftLeftLogicalSaturateUnsignedScalar: move scalar-op-from-vector-op code into shared functions

imhameed committed:
Fix PopCount

imhameed committed:
Fix ReverseElement8, ReverseElement16, ReverseElement32

imhameed committed:
Fix ExtractNarrowingSaturateScalar, ExtractNarrowingSaturateUnsignedScalar

imhameed committed:
More test fixes:

MultiplyDoublingSaturateHighScalar
MultiplyDoublingScalarBySelectedScalarSaturateHigh

MultiplyDoublingWideningSaturateScalarBySelectedScalar
MultiplyDoublingWideningScalarBySelectedScalarAndAddSaturate
MultiplyDoublingWideningScalarBySelectedScalarAndSubtractSaturate

MultiplyRoundedDoublingByScalarSaturateHigh
MultiplyRoundedDoublingBySelectedScalarSaturateHigh
MultiplyRoundedDoublingSaturateHighScalar
MultiplyRoundedDoublingScalarBySelectedScalarSaturateHigh
    - remove unnecessary special cases

MultiplyDoublingWideningSaturateScalar
    - add support for the special-case scalar LLVM intrinsic for sqdmull

imhameed committed:
LoadAndReplicateToVector: coerce the source pointer to the element type when loading a single element

imhameed committed:
Move OP_INSERT_* and OP_XCAST to a shared arm64/amd64 region

imhameed committed:
Address feedback: move IntrinsicId (and another LLVM-only anonymous enum) to a separate header

fanyang-mono reviewed (approved):
Thanks for this massive change!

imhameed committed:
Don't attempt to scalarize a non-scalar sqshlu

naricc commented on src/mono/mono/mini/mini-llvm.c:
```diff
@@ -303,6 +310,142 @@ static void create_aot_info_var (MonoLLVMModule *module);
 static void set_nonnull_load_flag (LLVMValueRef v);
 
+enum {
+	INTRIN_scalar = 1 << 0,
```
Is there any particular reason we are defining some of these with constant bit shifts, some with decimal literals, and some with hex literals?

imhameed committed:
MultiplyDoublingWideningSaturateScalar etc.: consistently place the scalar or scalar-in-vector return value in a Vector64

Remove OP_ARM64_ZERO_UPPER, which is unused

imhameed committed:
Explicitly zero out the unused bits in scalar ops built out of vector ops

undef can apparently pass through intrinsic functions during
optimization, so bias towards slightly worse but correct codegen for now

imhameed committed:
Fix the vector concatenation overloads of Vector128/256

imhameed committed:
Sha1.FixedRotate is a scalar-in-vector op. TODO: refactor to use XOP_SCALAR_X_X
